### PR TITLE
Bug 1414572 - Don't display, and evetually delete, expired entries

### PR DESCRIPTION
**Pull request by ydidwania**
Added `Namespace.expireEntries` to `src/data.js` .

### CODE REVIEW

**Comment by ydidwania** [comment]
Pinging @djmitche for review

File: src/data.js [hunk 1]
```diff
@@ -139,3 +139,40 @@ Namespace.ensureNamespace = function(namespace, expires) {
     });
   });
 };
+
+/**Delete expired entries */
+Namespace.expireEntries = function(parent, continuationToken) {
+  var that = this;
```
Comment (djmitche) [inline src/data.js]: Since you're using arrow functions (`=>`) below, this line should not be necessary.

File: src/data.js [hunk 2]
```diff
@@ -139,3 +139,40 @@ Namespace.ensureNamespace = function(namespace, expires) {
...
+  },
+  {
+    limit:         500,
+    continuation:   continuationToken,
+  }).then(async (data) => {
+    var dataLength = data.length;
+    // Stop recursion if the namespace is an indexed task
+    if (dataLength == 0) {
+      return;
+    }
```
Comment (djmitche) [inline src/data.js]: I think this conditional is unnecessary -- the loop below will not execute if `dataLength` is zero.  Also, sometimes Azure returns no results but a continuationToken (it has something to do with how the data is stored on their servers), in which case the `if (data.continuation)` below would evaluate as true.

File: src/data.js [hunk 3]
```diff
@@ -139,3 +139,40 @@ Namespace.ensureNamespace = function(namespace, expires) {
...
+      // entry without checking its children.
+      if (entry.expires.getTime() < Date.now()) {
+        entry.remove(false, true);
+      }
+    }            
+
+    if (data.continuation) {
+      await Namespace.expireEntries(parent, data.continuation) ;
+    }
+    return Promise.resolve(null);
```
Comment (djmitche) [inline src/data.js]: This is an async function, so there's no need to return anything here.

File: src/data.js (hunk 1 above)
Comment (ydidwania) [reply src/data.js]: We are using `that.query` in the next line. So will we not need that?

File: src/data.js [hunk 4]
```diff
@@ -139,3 +139,40 @@ Namespace.ensureNamespace = function(namespace, expires) {
...
+  var that = this;
+
+  return that.query({
+    parent: parent,
+  },
+  {
+    limit:         500,
+    continuation:   continuationToken,
+  }).then(async (data) => {
+    var dataLength = data.length;
```
Comment (ydidwania) [inline src/data.js]: Changed this to `dataLength = data.entries.length`

File: src/data.js (hunk 2 above)
Comment (ydidwania) [reply src/data.js]: OK. I removed that .

**Comment by ydidwania** [comment]
I wrote a pretty simple test which creates two tasks of which one is already expired and then run the expireEntries function. By putting various `console.log` statements, I found that the [query function](https://github.com/taskcluster/taskcluster-index/pull/33/files#diff-be99d06c129c1d2d7d05e3876e275e21R147) always returns `{ entries: [], continuation: null }
` . I am unable to find out why this  is happening

File: src/data.js (hunk 1 above)
Comment (djmitche) [reply src/data.js]: Not if you use `this.query` instead :)

**Comment by ydidwania** [comment]
We are only deleting [Namespaces](https://github.com/taskcluster/taskcluster-index/blob/master/src/data.js#L40) here while there is another table for  [IndexedTask](https://github.com/taskcluster/taskcluster-index/blob/master/src/data.js#L6) . Does deleting a namespace automatically delete the tasks  indexed within that namespace?

File: src/data.js (hunk 1 above)
Comment (ydidwania) [reply src/data.js]: oops

**Comment by ydidwania** [comment]
Pinging @djmitche for another review

**Comment by djmitche** [comment]
We're pretty busy at the Mozilla all-hands in Austin, TX right now, but this is on my list and I will look soon.

**Comment by ydidwania** [comment]
Its alright . I'll keep exploring other bugs in this time. 
Thanks and Good Luck for the various meetings and conferences !!

File: src/data.js [hunk 5]
```diff
@@ -141,33 +143,67 @@ Namespace.ensureNamespace = function(namespace, expires) {
...
       // from the root out to the leaf. A parent's expires is
       // always later than the child's. Hence, we can delete an
       // entry without checking its children.
+      console.log('expires ->', entry.expires.getTime());
+      console.log('current_time ->', Date.now());
       if (entry.expires.getTime() < Date.now()) {
         entry.remove(false, true);
       }
-    }            
+      await IndexedTask.ExpireTasks(namespace);
```
Comment (djmitche) [inline src/data.js]: I think this needs to happen before the entry itself is removed (where Namespace.expireEntries is called).

**Comment by ydidwania** [comment]
I have written the [test](https://pastebin.mozilla.org/9075222) in index_test.js , but since its not passing successfully, i haven't pushed it yet. Should I?

**Comment by djmitche** [comment]
Sure!  And if you'd like I can take a look and see if I can figure out why.

**Comment by ydidwania** [comment]
Can you please take a look at the function `expireEntries` and figure out why is the test not passing?

**Comment by djmitche** [comment]
That was a tough one!  I think the problem comes down to some "clever" stuff in azure-entities with dynamically creating subclasses.  The `Namespace` and `IndexedTask` module-level variables in `data.js` refer to a class that has not had its `setup` method called yet.  `main.js` calls the setup method, which creates a subclass.  Confusingly, it gives the same names to the subclass.

So the problem is that `Namespace.expireEntries` calls `IndexedTask.expireEntries`, and there `IndexedTask` has is the not-yet-setup class.  One of the things `setup` does is add the `__filterBuilder` method to the subclass it creates, and that `__filterBuilder` method is what `appendFilter` refers to.

If I comment out the call to `IndexedTask.expireEntires`, everything seems to work.

Could the solution be as simple as doing the expiration of the two tables entirely separately?  That is, to expire IndexedTasks, just scan for expired entries and remove them.  For Namespace, do the hierarchical scan as you've implemented it, but don't call `IndexedTasks.expireEntries`.

**Comment by ydidwania** [comment]
First of all thanks for debugging that and a clear explanantion. :):
I tried an alternate solution , passing `helper.handlers.IndexedTask` as a parameter to `expireEntries` in [index_test.js]()  and using that already set-up instance to call the `IndexedTask.expireTask` function, and it is working :): 
I am not sure if its correct to do so, should a function in Namespace class have an instance of another class as one of its parameters?

**Comment by ydidwania** [comment]
I will do the code cleanup once you approve of these changes :)

**Comment by djmitche** [comment]
I think this is OK.  Let's get it cleaned up and I'll test it out on the production data (with the removals replaced with logging the first time..)

**Comment by ydidwania** [comment]
Did that. Should I also add `debug` statements in the test?

**Comment by djmitche** [comment]
I pushed a change to log instead of removing.  So I'll land this now and see how it does.

**Comment by djmitche** [comment]
(the failed check is good -- it's showing that it didn't remove the task it should have!)

**Comment by djmitche** [comment]
OK, this is deployed.  We do need to add a section to `src/main.js`, but I'll take care of that.

**Comment by ydidwania** [comment]
Thanks for all the help on this one :)

**Comment by djmitche** [comment]
You did the hard work, I'm just deploying..

**Pull request merged**